doc/workshops/financial.rst

```diff
@@ -2,4 +2,721 @@
+
+.. code-block:: python
+
+   plt.figure(figsize=(16, 10))
+   fig = plt.scatter(port_1_vr[:, 0], port_1_vr[:, 1], c=port_1_sr, cmap="cool")
+   CB = plt.colorbar(fig)
+   CB.set_label("Sharpe ratio")
+   plt.xlabel("expected volatility")
+   plt.ylabel("expected return")
+   plt.title(" | ".join(SYMBOLS_1))
+
+You should see in the Plots pane something like this:
+
+.. image:: images/workshop-1/scatter-plot-sharpe-portfolio-1.png
+   :alt: Portfolio 1 Monte Carlo Sharpe simulation scatter plot
+
+A roughly linear relationship can be observed between returns and volatility: the higher the volatility, the higher the gains. And the Sharpe ratio shows an important amount of variability (it is noticeable in the "width" of the line drawn).
+
+This seems to be a good portfolio because it has a good performance with a not very large variance.
+
+
+Optimal portfolio weights
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Can we use the data obtained to calculate the optimal weights for the portfolio by year? Of course we can. Let's start by delimiting the previous years as variables.
+
+.. code-block:: python
+
+   start_year, end_year = (2012, 2020)
+
+We will now write a function to calculate these optimal weights.
+
+.. code-block:: python
+
+   def optimal_weights(returns, symbols, actual_weights, start_y, end_y):
+      
+       bounds = len(symbols) * [(0, 1), ]
+       constrains = {"type": "eq", "fun": lambda weights: weights.sum() - 1}
+       opt_weights = {}
+ 
+       for year in range(start_y, end_y):
+           _rets = returns[symbols].loc[f"{year}-01-01":f"{year}-12-31"]
+           _opt_w = minimize(lambda weights: -portfolio_sharpe(_rets, weights),
+                          actual_weights,
+                          bounds=bounds,
+                          constraints=constrains)["x"]
+           opt_weights[year] = _opt_w
+       return opt_weights
+       
+
+Let's describe this function in broad strokes. ``bounds`` indicates the maximum and minimum weights for each stock in the portfolio. The lowest weight will be 0 and the highest weight will be 1 for each stock in the portfolio. ``constrains`` is a function that ensures that the sum of the weights of all actions always adds up to 1. Then a loop is initialized that will segment the data for each year. In the variable ``_rets`` the returns for the specified year are obtained. In ``_opt_w`` the ``portfolio_shape()`` function is used to calculate the weights that maximize the Sharpe ratio. This is done with the ``minimize()`` function of Scipy (which takes as arguments the ``portfolio_shape`` function, the actual weights of our stocks in the portfolio, and the ``bounds`` and the ``constrains`` variables). Notice the - sign before ``portfolio_sharpe``? It's because ``minimize()`` aims to find the minimum value of a function relative to a parameter, but we are interested in the maximum, so we make the result of portfolio_sharpe a negative one.
```

Comment:
```suggestion
Let's describe this function in broad strokes. ``bounds`` indicates the maximum and minimum weights for each stock in the portfolio. The lowest weight will be 0 and the highest weight will be 1 for each stock in the portfolio. ``constrains`` is a function that ensures that the sum of the weights of all actions always adds up to 1. Then a loop is initialized that will segment the data for each year. In the variable ``_rets`` the returns for the specified year are obtained. In ``_opt_w`` the ``portfolio_shape()`` function is used to calculate the weights that maximize the Sharpe ratio. This is done with the ``minimize()`` function of Scipy, which takes as arguments the ``portfolio_shape`` function, the actual weights of our stocks in the portfolio, the ``bounds`` and the ``constrains`` variables. Did you notice the - sign before ``portfolio_sharpe``? It's because ``minimize()`` aims to find the minimum value of a function relative to a parameter, but we are interested in the maximum, so we make the result of portfolio_sharpe a negative one.
```